What's new for EF Core 6.0 (#3410)
dotnet/EntityFramework.Docs: entity-framework/core/what-is-new/ef-core-6.0/whatsnew.md

    .ToTable("Employees", b => b.IsTemporal());
-->
[!code-csharp[SimpleConfig](../../../../samples/core/Miscellaneous/NewInEFCore6/TemporalTablesSample.cs?name=SimpleConfig)]

When using EF Core to create the database, the new table will be configured as a temporal table with the SQL Server defaults for the timestamps and history table. For example, consider an `Employee` entity type:

<!--
    public class Employee
    {
        public Guid EmployeeId { get; set; }
        public string Name { get; set; }
        public string Position { get; set; }
        public string Department { get; set; }
        public string Address { get; set; }
        public decimal AnnualSalary { get; set; }
    }
-->
[!code-csharp[Employee](../../../../samples/core/Miscellaneous/NewInEFCore6/TemporalTablesSample.cs?name=Employee)]

The temporal table created will look like this:

```sql
DECLARE @historyTableSchema sysname = SCHEMA_NAME()
EXEC(N'CREATE TABLE [Employees] (
    [EmployeeId] uniqueidentifier NOT NULL,
    [Name] nvarchar(100) NULL,
    [Position] nvarchar(100) NULL,
    [Department] nvarchar(100) NULL,
    [Address] nvarchar(1024) NULL,
    [AnnualSalary] decimal(10,2) NOT NULL,

>     [PeriodEnd] datetime2 GENERATED ALWAYS AS ROW END NOT NULL,

Reviewer: HIIDDEN should be added to make them hidden?

Author: Good question! I will follow up.

Author: Filed https://github.com/dotnet/efcore/issues/25841

>     [PeriodStart] datetime2 GENERATED ALWAYS AS ROW START NOT NULL,

Reviewer: HIDDEN should be added for them to be hidden

    CONSTRAINT [PK_Employees] PRIMARY KEY ([EmployeeId]),
    PERIOD FOR SYSTEM_TIME([PeriodStart], [PeriodEnd])
) WITH (SYSTEM_VERSIONING = ON (HISTORY_TABLE = [' + @historyTableSchema + N'].[EmployeeHistory]))');
```

Notice that SQL Server creates two hidden `datetime2` columns called `PeriodEnd` and `PeriodStart`. These "period columns" represent the time range during which the data in the row existed. These columns are mapped to [shadow properties](xref:core/modeling/shadow-properties) in the EF Core model, allowing them to be used in queries as shown later.

> [!IMPORTANT]
> The times in these columns are always UTC time generated by SQL Server. UTC times are used for all operations involving temporal tables, such as in the queries shown below.

Notice also that an associated history table called `EmployeeHistory` is created automatically. The names of the period columns and history table can be changed with additional configuration to the model builder. For example:

> <!--

Author: It's intended to be quite detailed. Most of it will be redundant once we have the true docs, but that's okay because this captures a point-in-time and is tailored for people who want to know all about what's new EF Core 6.0, rather than what can be done with EF Core overall once 6.0 is released.

* Querying into the contents of primitive collections is not currently supported. Vote for [#16926](https://github.com/dotnet/efcore/issues/16926), [#25700](https://github.com/dotnet/efcore/issues/25700), and [#25701](https://github.com/dotnet/efcore/issues/25701) if these features are important to you.

### Translations to built-in functions

GitHub Issue: [#16143](https://github.com/dotnet/efcore/issues/16143).

The Cosmos provider now translates more Base Class Library (BCL) methods to Cosmos built-in-functions. The following tables show translations that are new in EF Core 6.0.

**String translations**

| BCL method          | Built-in function | Notes
|:--------------------|-------------------|------
| `String.Length`     | `LENGTH`          |
| `String.ToLower`    | `LOWER`           |
| `String.TrimStart`  | `LTRIM`           |
| `String.TrimEnd`    | `RTRIM`           |
| `String.Trim`       | `TRIM`            |
| `String.ToUpper`    | `UPPER`           |
| `String.Substring`  | `SUBSTRING`       |
| `+` operator        | `CONCAT`          |
| `String.IndexOf`    | `INDEX_OF`        |
| `String.Replace`    | `REPLACE`         |

> | `String.Equals`     | `STRINGEQUAL`     | Only case-insensitive calls

Author: Consider filing an issue if you think it's something we should do.

> | `String.Equals`     | `STRINGEQUAL`     | Only case-insensitive calls

Reviewer 2: https://github.com/dotnet/efcore/issues/25842

Logs the following diagnostic event:

```none
info: 8/30/2021 14:53:39.343 CosmosEventId.ExecutedReplaceItem[30105] (Microsoft.EntityFrameworkCore.Database.Command)
      Executed ReplaceItem (6 ms, 10.67 RU) ActivityId='1525b958-fea1-49e8-89f9-d429d0351fdb', Container='Shapes', Id='Impossible', Partition='TrianglesPartition'
```

Deleting an item from the Cosmos database generates the `CosmosEventId.ExecutedDeleteItem` event. For example, this code:

<!--
context.Remove(triangle);
context.SaveChanges();
-->
[!code-csharp[DeleteEvents](../../../../samples/core/Miscellaneous/NewInEFCore6.Cosmos/CosmosDiagnosticsSample.cs?name=DeleteEvents)]

Logs the following diagnostic event:

```none
info: 8/30/2021 14:53:39.359 CosmosEventId.ExecutedDeleteItem[30106] (Microsoft.EntityFrameworkCore.Database.Command)
      Executed DeleteItem (6 ms, 7.43 RU) ActivityId='cbc54463-405b-48e7-8c32-2c6502a4138f', Container='Shapes', Id='Impossible', Partition='TrianglesPartition'
```

### Configure throughput

GitHub Issue: [#17301](https://github.com/dotnet/efcore/issues/17301).

> The Cosmos model can now be configured with manual or auto-scale throughput. These values can be configured for all entity types at the model level. For example:

Reviewer 3: ```suggestion
The Cosmos model can now be configured with manual or auto-scale throughput. These values will provision throughput on the database. For example:
```

<!--
modelBuilder.HasManualThroughput(2000);
modelBuilder.HasAutoscaleThroughput(6000);
-->
[!code-csharp[ModelThroughput](../../../../samples/core/Miscellaneous/NewInEFCore6.Cosmos/CosmosModelConfigurationSample.cs?name=ModelThroughput)]

> In addition, individual entity types can be configured for specific throughput. For example:

Reviewer 3: ```suggestion
In addition, individual entity types can be configured to provision throughput for the corresponding container. For example:
```

<!--
modelBuilder.Entity<Family>(
    entityTypeBuilder =>
    {
        entityTypeBuilder.HasAutoscaleThroughput(3000);
        entityTypeBuilder.HasAutoscaleThroughput(12000);
    });
-->
[!code-csharp[EntityTypeThroughput](../../../../samples/core/Miscellaneous/NewInEFCore6.Cosmos/CosmosModelConfigurationSample.cs?name=EntityTypeThroughput)]

### Configure time-to-live

GitHub Issue: [#17307](https://github.com/dotnet/efcore/issues/17307).

> Entity types in the Cosmos model can now be configured with a default time-to-live, or a specific time-to-live for the analytical store. For example:

Reviewer 3: ```suggestion
Entity types in the Cosmos model can now be configured with the default time-to-live and time-to-live for the analytical store. For example:
```
